### DISCUSSION TITLE
Moving Buttons

### DISCUSSION THREAD

**Discussion by basten82** [kachurovskiy/nanoels#45] [Q&A]
Hello,
how is it possible to change the speed of the "Left and Right" buttons ?
For my old lathe it is to fast....

Thanks

**Comment by kachurovskiy** [comment]
Hi, try increasing the constant on line [21](https://github.com/kachurovskiy/nanoels/blob/main/h1/NanoEls.ino#L21) e.g. from

```
#define PULSE_MIN_US round(500 * (200.0 / MOTOR_STEPS)) // Microseconds to wait after high pulse, min.
```

to
```
#define PULSE_MIN_US round(1000 * (200.0 / MOTOR_STEPS)) // Microseconds to wait after high pulse, min.
```

**Comment by basten82** [comment]
Thanks for the information,

I changed to  "1000 ", and the carriage moves now a little slower. It is still to fast for me, so I changed to 2500 and it worked ok. But if I use normal feed with turning spindle the Steppermotor moves at higher pitch 1.0 mm/ rev further after stopping the spindle. Spindle and Stepper don't stop at the same time.

How can I fix this....?

Thanks

**Reply by kachurovskiy** [reply]
When you change PULSE_MIN_US, you limit the maximum speed of the stepper, which in turn lowers the max RPM of the spindle you can use for 1.0mm thread - so the stepper is still spinning, trying to catch up with the spindle even when the spindle has stopped turning.

There are 3 possible solutions here:

1. Increase the stepper max speed by lowering the PULSE_MIN_US param to 1500 or 1000
2. Use smaller pitch values e.g. 0.5mm
3. Use slower spindle speeds

**Comment by basten82** [comment]
Thanks for the Info.   

Could you please describe wich line you mean ?!  
For what is the  setting exactly in each line ? 

 ------ 
// Stepper pulse and acceleration constants. #define PULSE_MIN_US round(500 * (200.0 / MOTOR_STEPS)) // Microseconds to wait after high pulse, min. 

 #define PULSE_MAX_US round(2000 * (200.0 / MOTOR_STEPS)) // Microseconds to wait after high pulse, max. Slow start.
 --------  

Thanks

**Reply by kachurovskiy** [reply]
There are only 2 parameters relevant in the code for your case - MOTOR_STEPS and PULSE_MAX_US. But what really matters is that there's a fundamental problem with what you're asking for - low maximum stepper speed and running spindle at high RPM with high leadscrew pitch (1.0mm).

To be exact, if you want to limit your leadscrew RPM to 100 and it has a pitch of 2.0mm then setting ELS to 1.0mm you can't exceed 200 RPM on the spindle - no matter which ELS controller you use. If you exceed 200 RPM on the spindle, leadscrew falls behind and has to catch up after the spindle stops.

**Comment by basten82** [comment]
Hello thanks for the info, I got that. 
I would just like to know, what exactly these two lines do.  And what happens when I correct values up or down.
So please describe in detail. 

Positioning in rapid speed is not the best idea. No one do that at the lathe.
It would be nice if these 2 buttons could be controlled independently in contrast of the global max speed setting.
Then can set the user the slide exactly. Without the risk of crashing into the workpiece.

Thanks

**Reply by kachurovskiy** [reply]
`MOTOR_STEPS` lets the ELS know what micro-stepping is set on your stepper driver so that it can issue the correct number of step impulses to that driver.

Stepper driver moves the motor 1 step (360/MOTOR_STEPS degrees) when it receives a HIGH followed by LOW on the corresponding pin: https://github.com/kachurovskiy/nanoels/blob/main/h1/NanoEls.ino#L699

Key part of the logic here how long we wait between issuing those HIGH/LOW pulses e.g.:
- HIGH/LOW, wait 2ms, HIGH/LOW, wait 2ms, ... means the motor is moving rather slow
- HIGH/LOW, wait 1ms, HIGH/LOW, wait 1ms, ... means the motor is moving at 2x speed

Additional consideration here is that this "waiting" period between pulses can't change abruptly (e.g. from 1ms to 2ms right away) or the motor might stall, loose steps or be very noisy. It has to change gradually e.g. 1ms, 1.1ms, 1.2ms, ... 1.9ms, 2ms.

`PULSE_MAX_US` sets the distance between step pulses when the motor starts moving after being idle.

`PULSE_MIN_US` sets the minimum distance step pulses - so controls the maximum possible stepper speed.

> It would be nice if these 2 buttons could be controlled independently in contrast of the global max speed setting.
> Then can set the user the slide exactly. Without the risk of crashing into the workpiece.

That is a great point. I agree that it should be possible to precisely position the carriage using the move buttons, not just in pitch increments. We could re-purpose 1 pitch shortcut button for selecting the move scale (1x, 10x, 100x) in the ELS-off mode I guess? Because in ELS-on mode it has to move in pitch increments to stay in sync with the spindle.

**Comment by basten82** [comment]
Hello, 
thanks for the detailed description of the parameters. Now fine tuning is a little bit easier

**Is it possible to insert these parameters for "precisely position"  into the code ?!**

I think position in "ELS-OFF" mode with scaling 1x, 10x, 100x, is a good idea, but if this is to difficult to implement in the
code. Only one slow speed for the two buttons in the code is enough.

Thanks

**Reply by kachurovskiy** [reply]
The hard part about any change here is testing on hardware. I don't have the NanoEls attached to my lathe at the moment so it will be a while for this to be implemented.

Here is the proposal that might work to set different max speed for ON and OFF states - https://github.com/kachurovskiy/nanoels/compare/main...kachurovskiy-pulse-min-us?quick_pull=1 - please try this and let us know if it helps (didn't test it myself beyond compiling).

**Comment by basten82** [comment]
Hello, thanks for the code.

I tested the code today and it works. 
The motor now runs more slowly with the buttons in "ELS-On" mode.

But I think I can't go to very slow speed. Above a certain value, the stepper speed no longer changes ?!
Is there a dependency in the code ?!

Is it possible that the buttons only work in the slow speed ?

Thanks
